WordPress/openverse
# Alter Django Admin media views to surface content reports

## Fixes



Fixes #3635 by @sarayourfriend

## Description



This PR modifies the base media list views to accomplish the following:
- Exclude media items that have no reports
- Add columns for total report count, pending report count, oldest report date, and pending report links
- Automatically filter by records with pending reports, but allow viewing all reported records

This work was admittedly pretty hairy :sweat_smile: I'm hoping that the changes I've made here make sense, I'm certainly pleased that I was able to get all the functionality set up correctly! I'm happy to make anything more idiomatically "Django" where it makes sense to.

I also exposed the decision views so we could more easily add decisions for media, at least for testing. I'm sure that will change, and I'm honestly a little unclear on all of the moving parts here and how decisions will _actually_ be applied. This view currently links to each individual report, but since decisions can be applied to multiple reports _and_ multiple media records, I'm just not sure what all the actions will look like. That to say, this view might change a little once that's built out, which is okay!

### Screenshots
**Pending only filtered (default)**
![Screenshot 2024-05-01_17-01-53](https://github.com/WordPress/openverse/assets/10214785/4ba3b292-98e5-4e90-a9ff-946d5cdb3b13)

**All reported records**
![Screenshot 2024-05-01_17-02-02](https://github.com/WordPress/openverse/assets/10214785/661dbf7a-a4b8-40b1-847a-a3c71e3fb887)


## Testing Instructions


The easiest way to do this is the following:

1. Run `just down -v && just api/init`
2. Run `just api/dbshell` and execute the following:

```sql
COPY public.api_imagedecision (id, created_on, updated_on, notes, action, moderator_id) FROM stdin;
1	2024-05-01 23:27:09.884471+00	2024-05-01 23:27:09.884487+00		rejected_reports	3
2	2024-05-01 23:32:51.386486+00	2024-05-01 23:32:51.386501+00		marked_sensitive	3
\.


COPY public.api_imagedecision_media_objs (id, imagedecision_id, image_id) FROM stdin;
1	1	893
2	2	111
\.


COPY public.nsfw_reports (id, identifier, reason, description, status, created_at, decision_id) FROM stdin;
1	2e22c7e7-58b7-486c-9df8-16e84c27478f	mature		pending_review	2024-04-26 20:44:31.271278+00	\N
2	2e22c7e7-58b7-486c-9df8-16e84c27478f	dmca		pending_review	2024-04-26 20:44:36.178746+00	\N
4	2e22c7e7-58b7-486c-9df8-16e84c27478f	mature		pending_review	2024-04-26 20:44:46.869982+00	\N
5	c296b69b-c8d7-4fc3-b820-a83de6138e0e	dmca		pending_review	2024-04-26 20:44:52.034395+00	\N
6	de37812d-2daa-49f6-91ab-e87e780ca653	dmca	asdfsadfsadfsadfsadfsadf	pending_review	2024-04-26 20:45:31.272964+00	\N
7	de37812d-2daa-49f6-91ab-e87e780ca653	dmca		pending_review	2024-05-01 21:12:53.177529+00	\N
8	c296b69b-c8d7-4fc3-b820-a83de6138e0e	dmca		pending_review	2024-05-01 21:21:10.442789+00	\N
9	ba711b67-2cbe-487a-9feb-19d70accf912	dmca		pending_review	2024-05-01 21:28:45.869087+00	\N
3	2e22c7e7-58b7-486c-9df8-16e84c27478f	other		pending_review	2024-04-26 20:44:42.274724+00	1
10	052d2cfc-fdcb-40de-994c-40eadef753fc	dmca		pending_review	2024-05-01 21:31:04.020163+00	2
\.
```

Then you should be able to go to http://localhost:50280/admin/api/image/ and see the views in the screenshots above!

The actual `image_id`s in the `api_imagedecision_media_objs` table might be non-deterministic, I'm not sure. It makes me wonder too if we should alter that table to use `identifier` in its table reference rather than `{media}_id`. It's what we do for the sensitive/deleted objects tables as well: https://github.com/WordPress/openverse/blob/0db784ebc4d4937c925dcb14c1465bafb3e20c69/api/api/models/image.py#L97-L96. CC @dhruvkb as you might be the best informed on that.

## Checklist


      PRs) or the media properties generator (`just catalog/generate-docs media-props`
      for the catalog or `just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
